### PR TITLE
Portal WETH handling + more tests and refactoring

### PR DESCRIPTION
**Pull request by yarivbancor**

### CODE REVIEW

File: contracts/bancor-portal/BancorPortal.sol [hunk 1]
```diff
@@ -108,7 +111,8 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
         IUniswapV2Router02 uniswapV2Router,
         IUniswapV2Factory uniswapV2Factory,
         IUniswapV2Router02 sushiSwapV2Router,
-        IUniswapV2Factory sushiSwapV2Factory
+        IUniswapV2Factory sushiSwapV2Factory,
+        address weth
```
Comment (barakman) [inline contracts/bancor-portal/BancorPortal.sol:115]: Why not `IERC20`?

File: contracts/bancor-portal/BancorPortal.sol [hunk 2]
```diff
@@ -117,6 +121,7 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
         validAddress(address(uniswapV2Factory))
         validAddress(address(sushiSwapV2Router))
         validAddress(address(sushiSwapV2Factory))
+        validAddress(address(weth))
```
Comment (barakman) [inline contracts/bancor-portal/BancorPortal.sol:124]: As it stands, no need to cast `weth` to an address.

File: contracts/bancor-portal/BancorPortal.sol [hunk 3]
```diff
@@ -108,7 +111,8 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
         IUniswapV2Router02 uniswapV2Router,
```
Comment (barakman) [inline contracts/bancor-portal/BancorPortal.sol:111]: We typically name these input arguments: `initUniswapV2Router`, `initUniswapV2Factory`, etc.
This allows us to implement getter functions with the names that you are currently using, without collisions.

File: contracts/bancor-portal/BancorPortal.sol [hunk 4]
```diff
@@ -64,6 +64,9 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
     // SushiSwap v2 factory contract
     IUniswapV2Factory private immutable _sushiSwapV2Factory;
```
Comment (barakman) [inline contracts/bancor-portal/BancorPortal.sol:65]: Don't we want public getter functions for all of these private variables?

File: contracts/bancor-portal/BancorPortal.sol [hunk 5]
```diff
@@ -271,28 +277,32 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
         // look for relevant whitelisted pools, revert if there are none
         bool[2] memory whitelist;
```
Comment (barakman) [inline contracts/bancor-portal/BancorPortal.sol:278]: Do the conversion weth-to-native once (on declaration line), then you don't need to keep calling this function:
```
Token[2] memory tokens = [_wethToNative(token0), _wethToNative(token1)];
```

File: contracts/bancor-portal/BancorPortal.sol [hunk 6]
```diff
@@ -349,22 +359,47 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
...
         IERC20(address(pair)).safeApprove(address(router), poolTokenAmount);
-
         uint256 deadline = block.timestamp + MAX_DEADLINE;
-        if (tokens[0].isNative()) {
-            router.removeLiquidityETH(address(tokens[1]), poolTokenAmount, 1, 1, address(this), deadline);
-        } else if (tokens[1].isNative()) {
-            router.removeLiquidityETH(address(tokens[0]), poolTokenAmount, 1, 1, address(this), deadline);
+
+        Token token0 = _nativeToWETH(tokens[0]);
+        Token token1 = _nativeToWETH(tokens[1]);
```
Comment (barakman) [inline contracts/bancor-portal/BancorPortal.sol:365]: Can get rid of these two, and use `if (tokens[x].isNative())` below, where `x` is `0` or `1`.

**Review by barakman — APPROVED** [review]

File: contracts/bancor-portal/BancorPortal.sol (hunk 2 above)
Comment (yarivbancor) [reply contracts/bancor-portal/BancorPortal.sol]: outdated

File: contracts/bancor-portal/BancorPortal.sol (hunk 5 above)
Comment (yarivbancor) [reply contracts/bancor-portal/BancorPortal.sol:278]: done

File: contracts/bancor-portal/BancorPortal.sol (hunk 6 above)
Comment (yarivbancor) [reply contracts/bancor-portal/BancorPortal.sol]: done

File: contracts/bancor-portal/BancorPortal.sol [hunk 7]
```diff
@@ -105,33 +108,36 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
...
         _bnt = bnt;
-        _uniswapV2Router = uniswapV2Router;
-        _uniswapV2Factory = uniswapV2Factory;
-        _sushiSwapV2Router = sushiSwapV2Router;
-        _sushiSwapV2Factory = sushiSwapV2Factory;
+        _uniswapV2Router = initUniswapV2Router;
+        _uniswapV2Factory = initUniswapV2Factory;
+        _sushiSwapV2Router = initSushiSwapV2Router;
+        _sushiSwapV2Factory = initSushiSwapV2Factory;
+        _weth = initWETH;
```
Comment (lbeder) [inline contracts/bancor-portal/BancorPortal.sol:133]: You can also get it by calling `initUniswapV2Router.WETH()`. Might be slightly cleaner, since it won't require us to add another explicit argument (which is more error-prone since it can be accidentally 0x0 or different than `initUniswapV2Router.WETH()`.

Alternatively (or in addition), we can verify that `initWETH == initUniswapV2Router.WETH() == initSushiSwapV2Router.WETH()`. It is correct on mainnet but may require more tweaking on testnets or other chains, so it's better safe than sorry.

File: deploy/tests/000032-upgrade-bancor-portal-v2.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+        proxyAdmin = await DeployedContracts.ProxyAdmin.deployed();
+        bancorPortal = await DeployedContracts.BancorPortal.deployed();
+    });
+
+    it('should deploy and configure the bancor portal contract', async () => {
+        expect(await proxyAdmin.getProxyAdmin(bancorPortal.address)).to.equal(proxyAdmin.address);
+        expect(await bancorPortal.version()).to.equal(2);
+        await expectRoleMembers(bancorPortal, Roles.Upgradeable.ROLE_ADMIN, [deployer]);
+    });
+});
```
Comment (lbeder) [inline deploy/tests/000032-upgrade-bancor-portal-v2.ts:27]: We should add fork-only basic portal usage here and in `network.ts`. The test here is only supposed to do a quick sanity check that the upgraded portal works (e.g., by using it to migrate some positions), while the test in `network.ts` is the one that is going to be executed every 10 minutes and maybe more extensive.

IMHO, if you'll add a "good" test to `network.ts` (use the whales to create a both ETH and TKN positions for each protocol and then use the portal to migrate them to Bancor), then adding a similar test to `000032-upgrade-bancor-portal-v2.ts` **is not required**.

File: deployments/mainnet/types/factories/contracts/bancor-portal/BancorPortalV1__factory.ts [hunk 9]
```diff
@@ -0,0 +1,651 @@
+/* Autogenerated file. Do not edit manually. */
+
+/* tslint:disable */
+
+/* eslint-disable */
+import type {
+  BancorPortal,
+  BancorPortalInterface,
+} from "../../../contracts/bancor-portal/BancorPortal";
```
Comment (lbeder) [inline deployments/mainnet/types/factories/contracts/bancor-portal/BancorPortalV1__factory.ts:9]: You did everything right, other than this tricky (and currently WIP/not documented) change: it should be `"../../../contracts/bancor-portal/BancorPortalV1"` so that the factory references the deprecated type 😅 😅 😅

File: utils/TokenData.ts [hunk 10]
```diff
@@ -2,6 +2,7 @@ import { defaults } from 'lodash';
 
 export const DEFAULT_DECIMALS = 18;
 export const NATIVE_TOKEN_ADDRESS = '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE';
+export const WETH_ADDRESS = '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2';
```
Comment (lbeder) [inline utils/TokenData.ts:5]: We should only add symbolic/chain-agnostic data here, so `WETH_ADDRESS` should be either fetched from `data.ts` or used defined as a mock address in `test/bancor-portal/BancorPortal.ts` instead

File: test/bancor-portal/BancorPortal.ts [hunk 11]
```diff
@@ -740,4 +846,20 @@ describe('BancorPortal', () => {
...
+    const transfer = async (
+        sourceAccount: SignerWithAddress,
+        token: TokenWithAddress,
+        target: string | Addressable,
+        amount: BigNumberish
+    ) => {
+        const targetAddress = toAddress(target);
+        const tokenAddress = token.address;
+        if ([NATIVE_TOKEN_ADDRESS, WETH_ADDRESS].includes(tokenAddress)) {
+            return await sourceAccount.sendTransaction({ to: targetAddress, value: amount });
```
Comment (lbeder) [inline test/bancor-portal/BancorPortal.ts:859]: We usually don't `return await`- just `return` the promise (it's unneeded and node.js knows how to handle it)

**Review by lbeder — CHANGES_REQUESTED** [review]
Good stuff 👍

File: contracts/bancor-portal/BancorPortal.sol [hunk 12]
```diff
@@ -256,7 +262,7 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
         address provider
     ) private returns (MigrationResult memory) {
         // get Uniswap's pair
-        address pairAddress = factory.getPair(address(token0), address(token1));
+        address pairAddress = factory.getPair(address(_nativeToWETH(token0)), address(_nativeToWETH(token1)));
```
[the remark this answers is not in the document]
Comment (yarivbancor) [reply contracts/bancor-portal/BancorPortal.sol]: done

File: contracts/bancor-portal/BancorPortal.sol (hunk 7 above)
Comment (yarivbancor) [reply contracts/bancor-portal/BancorPortal.sol]: done

File: deployments/mainnet/types/factories/contracts/bancor-portal/BancorPortalV1__factory.ts (hunk 9 above)
Comment (yarivbancor) [reply deployments/mainnet/types/factories/contracts/bancor-portal/BancorPortalV1__factory.ts]: done

File: utils/TokenData.ts (hunk 10 above)
Comment (yarivbancor) [reply utils/TokenData.ts]: done

File: test/bancor-portal/BancorPortal.ts (hunk 11 above)
Comment (yarivbancor) [reply test/bancor-portal/BancorPortal.ts]: done

File: contracts/bancor-portal/BancorPortal.sol [hunk 13]
```diff
@@ -105,33 +108,34 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
         IBancorNetwork network,
```
Comment (barakman) [inline contracts/bancor-portal/BancorPortal.sol:108]: Well, I meant that we usually add 'init' to all construction arguments, but it is minor of course.

File: contracts/bancor-portal/BancorPortal.sol [hunk 14]
```diff
@@ -266,12 +270,13 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
...
         // arrange tokens in an array
-        Token[2] memory tokens = [token0, token1];
+        Token[2] memory tokens = [_wethToNative(token0), _wethToNative(token1)];
 
         // look for relevant whitelisted pools, revert if there are none
         bool[2] memory whitelist;
         for (uint256 i = 0; i < 2; i++) {
-            whitelist[i] = tokens[i].isEqual(_bnt) || _networkSettings.isTokenWhitelisted(tokens[i]);
+            Token token = tokens[i];
+            whitelist[i] = token.isEqual(_bnt) || _networkSettings.isTokenWhitelisted(token);
```
Comment (barakman) [inline contracts/bancor-portal/BancorPortal.sol:279]: You may as well revert this change.

File: contracts/bancor-portal/BancorPortal.sol [hunk 15]
```diff
@@ -285,14 +290,14 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
...
-            uint256 delta = tokens[i].balanceOf(address(this)) - previousBalances[i];
+            Token token = tokens[i];
+            uint256 delta = token.balanceOf(address(this)) - previousBalances[i];
             if (whitelist[i]) {
                 deposited[i] = delta;
-                _deposit(tokens[i], deposited[i], provider);
+                _deposit(token, deposited[i], provider);
             } else {
-                _transferToProvider(tokens[i], delta, provider);
+                _transferToProvider(token, delta, provider);
```
Comment (barakman) [inline contracts/bancor-portal/BancorPortal.sol:300]: You may as well revert these changes.

File: contracts/bancor-portal/BancorPortal.sol [hunk 16]
```diff
@@ -367,4 +372,34 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
...
+            return Token(address(_weth));
+        }
+        return token;
+    }
+
+    /**
+     * @dev replaces WETH with the native token
+     */
+    function _wethToNative(Token token) private view returns (Token) {
+        if (_isWETH(token)) {
```
Comment (barakman) [inline contracts/bancor-portal/BancorPortal.sol:390]: Function `_isWETH` seems redundant by now; you may as well do `if (address(token) == address(_weth))` here.

**Review by barakman — APPROVED** [review]

File: contracts/bancor-portal/BancorPortal.sol [hunk 17]
```diff
@@ -256,7 +260,7 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
         address provider
     ) private returns (MigrationResult memory) {
         // get Uniswap's pair
-        address pairAddress = factory.getPair(address(token0), address(token1));
+        address pairAddress = factory.getPair(address(_nativeToWeth(token0)), address(_nativeToWeth(token1)));
```
Comment (yudilevi) [inline contracts/bancor-portal/BancorPortal.sol:263]: we shouldn't call `_nativeToWeth` if it's not the native token - you could add a check at the beginning of the method and if it's indeed the native token, translate it to WETH right there.

File: contracts/bancor-portal/BancorPortal.sol [hunk 18]
```diff
@@ -266,12 +270,13 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
         Token(address(pair)).safeTransferFrom(provider, address(this), poolTokenAmount);
 
         // arrange tokens in an array
-        Token[2] memory tokens = [token0, token1];
+        Token[2] memory tokens = [_wethToNative(token0), _wethToNative(token1)];
```
Comment (yudilevi) [inline contracts/bancor-portal/BancorPortal.sol:273]: same as above, WETH functions should only be used on the native token.

File: contracts/bancor-portal/BancorPortal.sol (hunk 14 above)
Comment (yudilevi) [reply contracts/bancor-portal/BancorPortal.sol:279]: a bit more clear IMO

File: contracts/helpers/MockUniswapV2Pair.sol [hunk 19]
```diff
@@ -37,7 +38,7 @@ contract MockUniswapV2Pair is TestERC20Token, Utils {
         Token[2] memory tokens = [token0, token1];
 
         for (uint256 i = 0; i < 2; i++) {
-            if (tokens[i].isNative()) {
+            if (address(tokens[i]) == _weth) {
```
Comment (yudilevi) [inline contracts/helpers/MockUniswapV2Pair.sol:41]: is the contract ever going to hold ETH directly?

File: contracts/bancor-portal/BancorPortal.sol [hunk 20]
```diff
@@ -367,4 +372,34 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
             );
         }
     }
+
+    /**
+     * @dev replaces native token with WETH
+     */
+    function _nativeToWeth(Token token) private view returns (Token) {
```
Comment (yudilevi) [inline contracts/bancor-portal/BancorPortal.sol:379]: `_nativeToWeth` -> `_nativeToWETH`

File: test/bancor-portal/BancorPortal.ts [hunk 21]
```diff
@@ -422,11 +430,103 @@ describe('BancorPortal', () => {
...
+                    whitelistedToken1.address,
+                    whitelistedToken2.address,
+                    AMOUNT,
+                    AMOUNT,
+                    true,
+                    true
+                );
+        });
+
+        it('deposits when token1 is WETH and token2 is unlisted', async () => {
```
Comment (yudilevi) [inline test/bancor-portal/BancorPortal.ts:456]: unlisted -> not whitelisted

**Review by yudilevi — CHANGES_REQUESTED** [review]

File: contracts/bancor-portal/BancorPortal.sol [hunk 22]
```diff
@@ -367,4 +374,30 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
...
+        address token1Address = tokens[1].isNative() ? address(_weth) : address(tokens[1]);
+
+        address pairAddress = factory.getPair(token0Address, token1Address);
+        return IUniswapV2Pair(pairAddress);
+    }
+
+    /**
+     * @dev returns true if given token is WETH
+     */
+    function _isWETH(Token token) private view returns (bool) {
```
Comment (barakman) [inline contracts/bancor-portal/BancorPortal.sol:397]: `return address(token) == address(_weth);`

**Review by barakman — DISMISSED** [review]

File: contracts/bancor-portal/BancorPortal.sol [hunk 23]
```diff
@@ -285,14 +292,14 @@ contract BancorPortal is IBancorPortal, ReentrancyGuardUpgradeable, Utils, Upgra
...
-            uint256 delta = tokens[i].balanceOf(address(this)) - previousBalances[i];
+            Token token = tokens[i];
+            uint256 delta = token.balanceOf(address(this)) - previousBalances[i];
             if (whitelist[i]) {
                 deposited[i] = delta;
-                _deposit(tokens[i], deposited[i], provider);
+                _deposit(token, deposited[i], provider);
             } else {
-                _transferToProvider(tokens[i], delta, provider);
+                _transferToProvider(token, delta, provider);
```
Comment (yudilevi) [inline contracts/bancor-portal/BancorPortal.sol:302]: should _transferToProvider handle WETH unwrapping?

**Comment by lbeder** [comment]
> note that tests are failing

The tests have failed due to unrelated infra issues, so I'm rerunning them. FYI @yudilevi @yarivbancor

File: deploy/tests/000032-upgrade-bancor-portal-v2.ts (hunk 8 above)
Comment (lbeder) [reply deploy/tests/000032-upgrade-bancor-portal-v2.ts:27]: Resolving and will focus on `network.ts`

File: contracts/bancor-portal/BancorPortal.sol (hunk 23 above)
Comment (yudilevi) [reply contracts/bancor-portal/BancorPortal.sol:302]: cool

File: deploy/tests/network.ts [hunk 24]
```diff
@@ -396,5 +398,80 @@ import { getNamedAccounts } from 'hardhat';
                 }
             }
         });
+
+        it.skip('should perform portal migrations from UniswapV2 type pools', async () => {
```
Comment (yudilevi) [inline deploy/tests/network.ts:402]: skipped on purpose? failing?

Comment (lbeder) [reply deploy/tests/network.ts]: Updated and refactored

File: deploy/tests/network.ts [hunk 25]
```diff
@@ -623,131 +627,322 @@ import { getNamedAccounts } from 'hardhat';
...
 
         describe('migrations', () => {
-            context('with some deposits', () => {
+            describe('from v2', () => {
+                let anchor: Owned;
+                let converter: StandardPoolConverter;
+                let bnTKN: PoolToken;
+
                 beforeEach(async () => {
                     await stabilizePoolV3(NATIVE_TOKEN_ADDRESS, ethWhale);
```
Comment (yudilevi) [inline deploy/tests/network.ts:636]: unrelated but no need to stabilize before deposits

File: deploy/tests/network.ts [hunk 26]
```diff
@@ -623,131 +627,322 @@ import { getNamedAccounts } from 'hardhat';
...
+                        beforeEach(async () => {
+                            const { uniswapV2Router02, uniswapV2Factory, sushiSwapRouter, sushiSwapFactory } =
+                                await getNamedAccounts();
+
+                            await ethWhale.sendTransaction({
+                                value: NATIVE_TOKEN_DESIRED_AMOUNT,
+                                to: linkWhale.address
+                            });
+
+                            await stabilizePoolV3(NATIVE_TOKEN_ADDRESS, ethWhale);
```
Comment (yudilevi) [inline deploy/tests/network.ts:791]: not necessary to stabilize before depositing in v3

File: deploy/tests/network.ts [hunk 27]
```diff
@@ -623,131 +627,322 @@ import { getNamedAccounts } from 'hardhat';
...
+
+                            const { liquidity: newLinkLiquidity } = await poolCollection.poolData(link);
+
+                            expect(newLinkLiquidity.stakedBalance).to.be.almostEqual(
+                                prevLinkLiquidity.stakedBalance.add(depositedLinkAmount),
+                                {
+                                    maxRelativeError: new Decimal('0.0000000000000000001')
+                                }
+                            );
+                            expect({
```
Comment (yudilevi) [inline deploy/tests/network.ts:904]: trading liquidity might change if the pool doesn't have excess liquidity, maybe it's enough testing for staked amounts?

File: deploy/tests/network.ts (hunk 26 above)
Comment (lbeder) [reply deploy/tests/network.ts:791]: 👍

File: deploy/tests/network.ts (hunk 27 above)
Comment (yudilevi) [reply deploy/tests/network.ts:898]: sounds good

**Review by yudilevi — APPROVED** [review]
